README.md
# Developer Auth API

Utility functions for Developer Identity Management Service integration. This library provide the middleware to handle security into backend API validating requests against the service.

The package is developed with the aim of being used in conjunction with the rest of the packages of the platform, but it don't restrict use it as standalone package.

[![Version npm](https://img.shields.io/npm/v/@ecualead/auth.svg?style=flat-square)](https://www.npmjs.com/package/@ecualead/auth)[![npm Downloads](https://img.shields.io/npm/dm/@ecualead/auth.svg?style=flat-square)](https://npmcharts.com/compare/@ecualead/auth?minimal=true)

[![NPM](https://nodei.co/npm/@ecualead/auth.png?downloads=true&downloadRank=true)](https://nodei.co/npm/@ecualead/auth/)

## Installation

```bash
npm install @ecualead/auth
```

## Initialize the middleware

The authententication middleware must be initialized with the address of the Identity Management Service

```js
import { AuthenticationCtrl } from "@ecualead/auth";
AuthenticationCtrl.setup("https://myserver.com", true);
```

After the middleware is initialized we can use it to authenticate the API Backend or to validate a received request. Validations can only done over authorization with bearer tokens, in other cases always return not authorized.

The second parameter can be used to inject the access token to each request performed with the axios if the `Authorization` header isn't set. If you don't want to set the authentication you must set the header `noauth` with any value. If you prefer not use the interceptor you can set it to false or omit.

### API Backend authentication

When the API Backend is registered into the Identity Managemen Service a credentials account are generated with an id and a secret token. This values must be used to exchange them by an access token with the API scopes. The received access token can be used to perform requests agains other API Backends.

Backend authentication can be easily integrated with the cluster server initialization:

```js
import { ClusterServer } from "@ecualead/server";
import { AuthenticationCtrl } from "@ecualead/auth";

/**
 * Authenticate agains auth service
 */
function requestCredentials(): Promise<void> {
  return new Promise<void>((resolve) => {
    AuthenticationCtrl.setup("https://myserver.com", true);
    AuthenticationCtrl.authService("miserviceid", "myservicesecret")
      .catch((err) => {
        console.error('Invalid authentication');
      }).finally(() => {
        resolve();
      });
  });
}

/* Initialize cluster server */
const clusterServer = ClusterServer.setup(
  { running: requestCredentials },
);

/* Run cluster with base routes */
clusterServer.run({
  ...
});
```

In this case each slave worker will authenticate against the server. The given access token can be accesed directly to allow to use it to send request to external APIs that validate against the same Identity Management Service.

```js
const token: string = AuthenticationCtrl.token;
```

## Validate authentication middleware

To authenticate the recived request we must use the middleware inside the expres router

```js
import { Router, Request, Response, NextFunction } from "express";
import { AuthenticationCtrl, SCOPE_VALIDATION } from "@ecualead/auth";
const router = Router();

router.get(
  "/hello",
  AuthenticationCtrl.middleware(["user", "admin"], SCOPE_VALIDATION.AND),
  (req: Request, res: Response, next: NextFunction) => {
    res.send("Hello World");
    res.end();
  }
);

export default router;
```

In this case we protect the request. This request can be performed only by an user that holds the `user` and `admin` scopes. If the user don't holds both scopes then can't access the resource.

```js
AuthenticationCtrl.middleware(scope?: string | string[],validation?: SCOPE_VALIDATION)
```

In case of array of scopes, it can be validated using one of the following three operations.

```js
enum SCOPE_VALIDATION {
  AND = 1, // Must contain all scopes
  OR = 2, // Must contain at least one scope
  NOT = 3 // Must not contain any scope
}
```

Once the request is authenticated, there is some information that can be used into the backedn api to handle the service data. After success authentication we can receive the user, application, project, domain or module that is doing the request and all the scopes authorized for that request.

```js
interface IAuthentication {
  user: string;
  application: string;
  project: string;
  domain: string;
  module: string;
  scope: string[];
}
```

This information can be accessed for the express response o request:

```js
req["user"];
res.locals["auth"];
```

If the `get` request can't sent the authorization token in the header request, the token can be taken from query parameters also, using the  following middlewares combination:

```js
import { Router, Request, Response, NextFunction } from "express";
import { AuthenticationCtrl, SCOPE_VALIDATION } from "@ecualead/auth";
const router = Router();

router.get(
  "/hello",
  AuthenticationCtrl.forceAuthToken("token"),
  AuthenticationCtrl.middleware(["user", "admin"], SCOPE_VALIDATION.AND),
  (req: Request, res: Response, next: NextFunction) => {
    res.send("Hello World");
    res.end();
  }
);

export default router;
```

In this case the middleware receive the access token into the `token` query parameter. By default middleware parameter can be omited and `token` is set by default, or you can change the field name.


CHANGELOG.md
# CHANGELOG

## [2.1.13] -2022-02-28

- Fix GoogleReCaptcha validation
- Validate recaptcha action

## [2.1.12] - 2022-01-31

- Initial GoogleReCaptcha v3 support

## [2.1.11] - 2022-01-12

- Add function to set initial token

## [2.1.10] - 2022-01-11

- Update errors definition

## [2.1.9] - 2022-01-11

- Update authentication middleware
- Add support for service internal security

## [2.1.7] - 2022-01-09

- Return JWT decoded token in middlewares

## [2.1.5] - 2022-01-07

- Remove duplicated interfaces

## [2.1.4] - 2022-01-07

- Fix JWT input data

## [2.1.3] - 2022-01-07

- Add support for JWT controllers
- Add predefined constants

## [2.1.1] - 2022-01-04

- Fix data export

## [2.1.0] - 2022-01-03

- Update package dependencies
- Allow to validate JWT

## [2.0.3] - 2021-10-13

- Fix error responses: Add `data` field return to error

## [2.0.2] - 2021-09-26

- Update package dependencies

## [2.0.1] - 2021-08-04

- Fix source code merge with version 1.1.4
- Fix changelog date entry for version 2.0.0

## [2.0.0] - 2021-08-03

- Update error handler schema
- Update to @ikoabo/server version 2.0.0

## [1.1.4] - 2021-07-18

- Add confirm action
- Add resend action
- Add recover actions
- Add validation action

## [1.1.3] - 2021-07-18

- Fix credential utility class export

## [1.1.2] - 2021-07-18

- Addings header authorization credential parser

## [1.1.1] - 2021-07-18

- Addings auth server actions controller
- Refactor internal source code
- Adding data types for auth handling

## [1.1.0] - 2021-07-17

- Update errors definition
- Fix url to login
- Handle project in auth urls

## [1.0.19] - 2020-10-19

- Adding query bearer token authentication if no header present

## [1.0.18] - 2020-10-13

- Return username on authentication validation

## [1.0.17] - 2020-09-04

- Add middleware to allow pass access token into query parameters

## [1.0.16] - 2020-09-03

- Fix scope AND validation
- Replace request dependencie by axios
- Add axios interceptor
- Update dependencies

## [1.0.14] - 2020-08-23

- Initial public version
- Predefined authentication error constants
- Authentication middleware
- Authentication scope validation


src/api/Authenticator.ts
/**
 * @Author: Reinier Millo Sánchez <millo>
 * @Date:   2020-04-12T21:28:26-05:00
 * @Email:  reinier.millo88@gmail.com
 * @Project: IKOABO Auth Microservice API
 * @Filename: Authenticator.ts
 * @Last modified by:   millo
 * @Last modified time: 2020-05-11T02:17:57-05:00
 * @Copyright: Copyright 2020 IKOA Business Opportunity
 */

import request from 'request';
import { Request, Response, NextFunction } from 'express';
import { Objects, HTTP_STATUS, Logger } from '@ikoabo/core_srv';
import { ERRORS } from '../types/errors';

export interface IAuthInfo {
  user: string;
  application: string;
  project: string;
  domain: string;
  scope: string[];
};

export class Authenticator {
  private static _instance: Authenticator;
  private _logger: Logger;
  private _authService: string;
  private _token: string;
  private _retries: number;

  /**
   * Private constructor to allow singleton instance
   */
  private constructor() {
    this._logger = new Logger('Authenticator');
    this._retries = 0;
  }

  /**
   * Retrieve singleton class instance
   */
  public static get shared(): Authenticator {
    if (!Authenticator._instance) {
      Authenticator._instance = new Authenticator();
    }
    return Authenticator._instance;
  }

  /**
   * Set the auth service url
   *
   * @param authService  Auth service url to use
   */
  public setup(authService: string) {
    this._authService = authService;
  }

  /**
   * Authenticate against the auth server using the given token and validate the scope
   *
   * @param token  Token to authenticate agains auth server
   * @param scope  Scope to validate
   */
  authenticate(token: string, scope?: string | string[]): Promise<IAuthInfo> {
    const self = Authenticator.shared;
    return new Promise<IAuthInfo>((resolve, reject) => {
      if (!self._authService || self._authService.length <= 0) {
        this._logger.error('Invalid auth service configuration');
        reject({ boError: ERRORS.INVALID_AUTH_SERVER, boStatus: HTTP_STATUS.HTTP_INTERNAL_SERVER_ERROR });
        return;
      }

      if (!token) {
        reject({ boError: ERRORS.INVALID_TOKEN });
        return;
      }

      /* Prepare the request options with the given token */
      let opts = {
        auth: { bearer: token },
      };

      /* Validate the token against the auth service */
      request.post(`${self._authService}/v1/oauth/authenticate`, opts, (error: any, response: request.Response, body: any) => {
        if (error) {
          this._logger.error('Invalid auth server response ', error);
          reject({ boError: ERRORS.UNKNOWN_AUTH_SERVER_ERROR, boStatus: HTTP_STATUS.HTTP_INTERNAL_SERVER_ERROR });
          return;
        }
        try { /* Try to convert response body to JSON */
          body = JSON.parse(body);
        } catch {
          reject({ boError: ERRORS.UNKNOWN_AUTH_SERVER_ERROR, boStatus: HTTP_STATUS.HTTP_INTERNAL_SERVER_ERROR });
          return;
        }

        /* Check for success/error response */
        if (body['error']) {
          reject({ boStatus: response.statusCode, boError: body['error'], boData: body['data'] });
          return;
        }

        /* On success prepare the response information */
        let auth: IAuthInfo = {
          user: Objects.get(body, 'user', null),
          application: Objects.get(body, 'application', null),
          project: Objects.get(body, 'project', null),
          domain: Objects.get(body, 'domain', null),
          scope: Objects.get(body, 'scope', []),
        }

        /* Check for valid response data */
        if (!auth.domain || !auth.project || !auth.application) {
          reject({ boError: ERRORS.AUTHENTICATION_REQUIRED, boStatus: HTTP_STATUS.HTTP_UNAUTHORIZED });
          return;
        }

        /* Validate the user scopes */
        this.validateScope(auth, scope).then(resolve).catch(reject);
      });
    });
  }

  /**
   * Validate the scope with the authentication information
   *
   * @param auth  Authentication information
   * @param scope  Scope to be validated
   */
  public validateScope(auth: IAuthInfo, scope?: string | string[]): Promise<IAuthInfo> {
    return new Promise<IAuthInfo>((resolve, reject) => {
      /* Validate required scopes */
      if (typeof scope === 'string') {
        if (auth.scope.indexOf(scope) < 0) {
          reject({ boError: ERRORS.INVALID_SCOPE, boStatus: HTTP_STATUS.HTTP_FORBIDDEN });
          return;
        }
      } else if (Array.isArray(scope)) {
        /* Scope is an array with multiple scopes. User must hold any of the given scope */
        if (scope.filter(value => auth.scope.indexOf(value) >= 0).length === 0) {
          reject({ boError: ERRORS.INVALID_SCOPE, boStatus: HTTP_STATUS.HTTP_FORBIDDEN });
          return;
        }
      } else if (scope) {
        reject({ boError: ERRORS.INVALID_SCOPE, boStatus: HTTP_STATUS.HTTP_FORBIDDEN });
        return;
      }

      resolve(auth);
    });
  }

  /**
   * Express middleware to authenticate the user request
   *
   * @params scope  Scope to be validated for the given user or application
   */
  middleware(scope?: string | string[]): (req: Request, res: Response, next: NextFunction) => void {
    const self = Authenticator.shared;
    return (req: Request, res: Response, next: NextFunction) => {
      /* Get authorization header token */
      const header: string[] = req.headers.authorization ? req.headers.authorization.split(' ') : [];
      if (!header || header.length !== 2) {
        next({ boError: ERRORS.AUTHENTICATION_REQUIRED, boStatus: HTTP_STATUS.HTTP_UNAUTHORIZED });
        return;
      }

      /* Check if the request was authenticated previously */
      if (res.locals['auth']) {
        self.validateScope(res.locals['auth'], scope)
          .then(() => {
            next();
          }).catch(next);
        return;
      }

      /* Authenticate the current request */
      self.authenticate(header[1], scope)
        .then((auth: IAuthInfo) => {
          res.locals['auth'] = auth;
          (<any>req)['user'] = auth;
          next();
        }).catch(next);
    };
  }

  /**
   * Authenticate the service application
   *
   * @param id  Application id of the service
   * @param secret  Application secret of the service
   */
  public authService(id: string, secret: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      /* Validate the auth service configuration */
      if (!this._authService || this._authService.length <= 0 || !id || !secret || id.length === 0 || secret.length === 0) {
        this._logger.error('Invalid auth service application authentication');
        reject({ boError: ERRORS.INVALID_AUTH_SERVER, boStatus: HTTP_STATUS.HTTP_INTERNAL_SERVER_ERROR });
        return;
      }

      /* Prepare the authentication credentials */
      let opts = {
        auth: { username: id, password: secret },
        form: {
          grant_type: 'client_credentials',
        },
      };

      this._retries++;
      this._logger.debug('Authenticating service', { retry: this._retries });

      /* Perform the authentication request against the IAM */
      request.post(`${this._authService}/v1/oauth/signin`, opts, (error: any, response: request.Response, body: any) => {
        /* Reject on error */
        if (error) {
          this._logger.error('Invalid auth server response ', error);
          /* Retry the request after 1 second */
          setTimeout(() => {
            this.authService(id, secret)
              .then(() => {
                resolve();
              }).catch(reject);
          }, 1000);
          return;
        }

        /* Try to convert response body to JSON */
        try {
          body = JSON.parse(body);
        } catch (err) {
          this._logger.error('Invalid auth server response ', err);
          /* Retry the request after 1 second */
          setTimeout(() => {
            this.authService(id, secret)
              .then(() => {
                resolve();
              }).catch(reject);
          }, 1000);
          return;
        }

        /* Check for success/error response */
        if (body['error']) {
          reject({ boStatus: response.statusCode, boError: body['error'], boData: body['data'] });
          return;
        }

        this._logger.debug('Service account authenticated');
        this._token = Objects.get(body, 'accessToken', null);
        resolve();
      });
    });
  }

  /**
   * Retrieve the auth token
   */
  public get token(): string {
    return this._token;
  }
}
